File: Data Structures & Algorithms/search-for-word-ii/submission-2.cpp
class Solution {
    struct Node {
        bool end;
        Node* next[26];
        Node() {
            this->end = false;
            for(int i = 0; i < 26; ++i) {
                this->next[i] = nullptr;
            }
        }
    };
    Node root;
public:
    void addWord(string& word) {
        Node* curr = &this->root;
        for(auto& c: word) {
            if(!curr->next[int(c) - int('a')]) {
                curr->next[int(c) - int('a')] = new Node();
            }
            curr = curr->next[int(c) - int('a')];
        }
        curr->end = true;
    }
    void addWords(vector<string>& words) {
        for(auto& word: words) {
            addWord(word);
        }
    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        addWords(words);
        vector<string> res;
        for(int i = 0; i < board.size(); ++i) {
            for(int j = 0; j < board[0].size(); ++j) {
                search(res, board, i, j, "", &this->root, {});
            }
        }
        return res;
    }
    void search(vector<string>& res, vector<vector<char>>& board, int r, int c, string currWord, Node* currNode, unordered_set<string> visited) {
        // cout << r << " " << c << " " << currWord << " " << currNode->end << endl;        
        if(r >= board.size() || c >= board[0].size() || visited.count(to_string(r) + "#" + to_string(c))) {
            return;
        }

        if(!currNode->next[int(board[r][c]) - int('a')]) {
            return;
        }
        currWord += board[r][c];
        currNode = currNode->next[int(board[r][c]) - int('a')];
        if(currNode->end == true) {
            res.push_back(currWord);
            currNode->end = false;
        }
        visited.insert(to_string(r) + "#" + to_string(c));
        search(res, board, r + 1, c, currWord, currNode, visited);
        search(res, board, r, c + 1, currWord, currNode, visited);
        search(res, board, r - 1, c, currWord, currNode, visited);
        search(res, board, r, c - 1, currWord, currNode, visited);
    }
};

File: Data Structures & Algorithms/design-word-search-data-structure/submission-2.py
class Node:
    def __init__(self):
        self.end = False
        self.right = [None] * 26

class WordDictionary:

    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        curr = self.root

        for c in word:
            if not curr.right[ord(c) - ord('a')]:
                curr.right[ord(c) - ord('a')] = Node()
            curr = curr.right[ord(c) - ord('a')]

        curr.end = True

    def search_helper(self, curr, word, i):
        if curr == None:
            return False

        if i == len(word):
            return curr.end

        res = False
        if word[i] == '.':
            for j in range(26):
                res = res or self.search_helper(curr.right[j], word, i + 1)

            return res
        else:
            curr = curr.right[ord(word[i]) - ord('a')]
            return self.search_helper(curr, word, i + 1)

    def search(self, word: str) -> bool:
        if len(word) == 0:
            return True

        return self.search_helper(self.root, word, 0)

        
